xblockutils/studio_editable.py

```diff
@@ -0,0 +1,295 @@
+        if info["type"] == "generic":
+            # Convert value to JSON string if we're treating this field generically:
+            info["value"] = json.dumps(info["value"])
+            info["default"] = json.dumps(info["default"])
+        if 'values_provider' in field.runtime_options:
+            values = field.runtime_options["values_provider"](self)
+        else:
+            values = field.values
+        if values and not isinstance(field, Boolean):
+            info['has_values'] = True
+            # This field has only a limited number of pre-defined options.
+            # Protip: when defining the field, values= can be a callable.
+            if isinstance(field.values, dict) and isinstance(field, (Float, Integer)):
+                # e.g. {"min": 0 , "max": 10, "step": .1}
+                info["min"] = field.values["min"]
+                info["max"] = field.values["max"]
+                info["step"] = field.values["step"]
+            else:
+                # e.g. [1, 2, 3] or [ {"display_name": "Always", "value": "always"}, {...}, ... ]
+                if not isinstance(values[0], dict) or "display_name" not in values[0]:
+                    values = [{"display_name": val, "value": val} for val in values]
+                info['values'] = values
+        if info["type"] in ("list", "set") and field.runtime_options.get('list_values_provider'):
+            list_values = field.runtime_options['list_values_provider'](self)
+            # list_values must be a list of values or {"display_name": x, "value": y} objects
+            # Furthermore, we need to convert all values to JSON since they could be of any type
+            if list_values and (not isinstance(list_values[0], dict) or "display_name" not in list_values[0]):
+                list_values = [json.dumps(val) for val in list_values]
+                list_values = [{"display_name": val, "value": val} for val in list_values]
+            else:
+                for entry in list_values:
+                    entry["value"] = json.dumps(entry["value"])
+            info['list_values'] = list_values
+            info['has_list_values'] = True
+        return info
+
+    @XBlock.json_handler
+    def submit_studio_edits(self, data, suffix=''):
+        """
+        AJAX handler for studio_view() Save button
+        """
+        values = {}  # dict of new field values we are updating
+        to_reset = []  # list of field names to delete from this XBlock
+        for field_name in self.editable_fields:
+            field = self.fields[field_name]
+            if field_name in data['values']:
+                if isinstance(field, JSONField):
+                    values[field_name] = field.from_json(data['values'][field_name])
+                else:
+                    raise JsonHandlerError(400, "Unsupported field type: {}".format(field_name))
+            elif field_name in data['defaults'] and field.is_set_on(self):
```

Comment:
Not necessarily. While it's true that the accompanying JavaScript does always send either an explicit value or a default value for every field, there's no reason to require that every editable field is always edited at the same time in our AJAX handler here. Subclasses or other potential uses of this AJAX method may need a way to say "make no change to field x", and they could do so by leaving that field out.